sklearn/tree/_criterion.pyx

```diff
@@ -962,6 +969,304 @@ cdef class MSE(RegressionCriterion):
 
+cdef class MAE(RegressionCriterion):
+    """Mean absolute error impurity criterion"""
+    def __dealloc__(self):
+        """Destructor."""
+
+        free(self.node_medians)
+
+    cdef np.ndarray left_child_heaps
+    cdef np.ndarray right_child_heaps
+    cdef double* node_medians
+
+    def __cinit__(self, SIZE_t n_outputs):
+        """Initialize parameters for this criterion.
+
+        Parameters
+        ----------
+        n_outputs: SIZE_t
+            The number of targets to be predicted
+        """
+
+        # Default values
+        self.y = NULL
+        self.y_stride = 0
+        self.sample_weight = NULL
+
+        self.samples = NULL
+        self.start = 0
+        self.pos = 0
+        self.end = 0
+
+        self.n_outputs = n_outputs
+        self.n_node_samples = 0
+        self.weighted_n_node_samples = 0.0
+        self.weighted_n_left = 0.0
+        self.weighted_n_right = 0.0
+
+        # Allocate accumulators. Make sure they are NULL, not uninitialized,
+        # before an exception can be raised (which triggers __dealloc__).
+        self.node_medians = NULL
+
+        # Allocate memory for the accumulators
+        self.node_medians = <double*> calloc(n_outputs, sizeof(double))
+
+        if (self.node_medians == NULL):
+            raise MemoryError()
+
+        self.left_child_heaps = np.empty(n_outputs, dtype='object')
+        self.right_child_heaps = np.empty(n_outputs, dtype='object')
+
+    cdef void init(self, DOUBLE_t* y, SIZE_t y_stride, DOUBLE_t* sample_weight,
+                   double weighted_n_samples, SIZE_t* samples, SIZE_t start,
+                   SIZE_t end) nogil:
+        """Initialize the criterion at node samples[start:end] and
+           children samples[start:start] and samples[start:end]."""
+
+        # Initialize fields
+        self.y = y
+        self.y_stride = y_stride
+        self.sample_weight = sample_weight
+        self.samples = samples
+        self.start = start
+        self.end = end
+        self.n_node_samples = end - start
+        self.weighted_n_samples = weighted_n_samples
+        self.weighted_n_node_samples = 0.
+
+        cdef SIZE_t i
+        cdef SIZE_t p
+        cdef SIZE_t k
+        cdef DOUBLE_t y_ik
+        cdef DOUBLE_t w_y_ik
+        cdef DOUBLE_t w = 1.0
+
+        # Fill accumulators with MedianHeaps
+        with gil:
+            for k in range(self.n_outputs):
+                self.left_child_heaps[k] = MedianHeap(self.n_node_samples)
+                self.right_child_heaps[k] = MedianHeap(self.n_node_samples)
+
+        cdef void** left_child_heaps = <void**> self.left_child_heaps.data
+        cdef void** right_child_heaps = <void**> self.right_child_heaps.data
+
+        for p in range(start, end):
+            i = samples[p]
+
+            if sample_weight != NULL:
+                w = sample_weight[i]
+
+            for k in range(self.n_outputs):
+                y_ik = y[i * y_stride + k]
+                w_y_ik = w * y_ik
```

Comment:
Yeah, raghav and I talked for quite a bit about this and we thought that
this would be correct because mse did it. I see your point, though, and I
think you're correct. How do you think we could extend the median heap to
use weights?

On Thu, Jun 23, 2016, 3:36 PM Gilles Louppe <notifications@github.com>
wrote:

> In sklearn/tree/_criterion.pyx
> <https://github.com/scikit-learn/scikit-learn/pull/6667#discussion_r68188020>
> :
>
> > +            for k in range(self.n_outputs):
> > +                self.left_child_heaps[k] = MedianHeap(self.n_node_samples)
> > +                self.right_child_heaps[k] = MedianHeap(self.n_node_
>
> samples)
> > +
> > +        cdef void** left_child_heaps = <void**> self.left_child_heaps.data
> > +        cdef void** right_child_heaps = <void**> self.right_child_heaps.data
> > +
> > +        for p in range(start, end):
> > +            i = samples[p]
> > +
> > +            if sample_weight != NULL:
> > +                w = sample_weight[i]
> > +
> > +            for k in range(self.n_outputs):
> > +                y_ik = y[i * y_stride + k]
> > +                w_y_ik = w * y_ik
>
> Hmmm, is this really correct? Shall you really push w * y_ik into the
> heaps to compute the median? The weight is not a scaling factor of the
> value itself, but of the number of times it appears. For example, if w ==
> 2.0, it is semantically equivalent to having y_ik appearing twice, but
> not as having 2.0 * y_ik.
>
> This should be fixed, as it is otherwise flawing the whole implementation.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/6667/files/20d6107c7ebfba1b96d483cfacb97a9ac304b3cb#r68188020>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AG72X-hkY4SxM0FTRuwVq2vpRpvmC3quks5qOjeLgaJpZM4IIwnV>
> .
>